frame/staking/src/migrations.rs

```diff
@@ -22,45 +22,79 @@ use frame_support::traits::OnRuntimeUpgrade;
+impl<T: Config> OnRuntimeUpgrade for InjectValidatorsIntoVoterList<T> {
 	fn on_runtime_upgrade() -> Weight {
 		if StorageVersion::<T>::get() == Releases::V8_0_0 {
 			for (v, _) in Validators::<T>::iter() {
 				let weight = Pallet::<T>::vote_weight(&v);
-				let _ = T::SortedListProvider::on_insert(v.clone(), weight).map_err(|err| {
-					log!(warn, "failed to insert {:?} into SortedListProvider: {:?}", v, err)
+				let _ = T::VoterList::on_insert(v.clone(), weight).map_err(|err| {
+					log!(warn, "failed to insert {:?} into VoterList: {:?}", v, err)
 				});
 			}
 
+			StorageVersion::<T>::put(Releases::V9_0_0);
 			T::BlockWeights::get().max_block
 		} else {
-			log!(warn, "InjectValidatorsIntoSortedListProvider being executed on the wrong storage version, expected Releases::V8_0_0");
+			log!(warn, "InjectValidatorsIntoVoterList being executed on the wrong storage version, expected Releases::V8_0_0");
 			T::DbWeight::get().reads(1)
 		}
 	}
 
 	#[cfg(feature = "try-runtime")]
 	fn pre_upgrade() -> Result<(), &'static str> {
 		use frame_support::traits::OnRuntimeUpgradeHelpersExt;
-		frame_support::ensure!(
-			StorageVersion::<T>::get() == crate::Releases::V8_0_0,
-			"must upgrade linearly"
-		);
+		ensure!(StorageVersion::<T>::get() == crate::Releases::V8_0_0, "must upgrade linearly");
 
-		let prev_count = T::SortedListProvider::count();
+		let prev_count = T::VoterList::count();
 		Self::set_temp_storage(prev_count, "prev");
 		Ok(())
 	}
 
 	#[cfg(feature = "try-runtime")]
 	fn post_upgrade() -> Result<(), &'static str> {
 		use frame_support::traits::OnRuntimeUpgradeHelpersExt;
-		let post_count = T::SortedListProvider::count();
+		let post_count = T::VoterList::count();
 		let prev_count = Self::get_temp_storage::<u32>("prev").unwrap();
 		let validators = Validators::<T>::count();
-		assert!(post_count == prev_count + validators);
+		ensure!(post_count == prev_count + validators, "incorrect count");
+		ensure!(StorageVersion::<T>::get(), Releases::V9_0_0, "version not set");
+		Ok(())
+	}
+}
+
+/// Migration implementation that injects all validators into sorted list.
+///
+/// This is only useful for chains that started their `VoterList` just based on nominators.
+pub struct InjectValidatorsIntoTargetList<T>(sp_std::marker::PhantomData<T>);
+impl<T: Config> OnRuntimeUpgrade for InjectValidatorsIntoTargetList<T> {
+	fn on_runtime_upgrade() -> Weight {
+		if StorageVersion::<T>::get() == Releases::V9_0_0 {
+			for (v, _) in Validators::<T>::iter() {
+				let weight = Pallet::<T>::vote_weight(&v);
+				let _ = T::TargetList::on_insert(v.clone(), weight).map_err(|err| {
+					log!(warn, "failed to insert {:?} into TargetList: {:?}", v, err)
+				});
+			}
+
+			StorageVersion::<T>::put(Releases::V10_0_0);
+			T::BlockWeights::get().max_block
+		} else {
+			log!(warn, "InjectValidatorsIntoTargetList being executed on the wrong storage version, expected Releases::V9_0_0");
```

Comment:
```suggestion
			log!(warn, "InjectValidatorsIntoTargetList being attempted on the wrong storage version, expected Releases::V9_0_0");
```
Same rationale as above - although not really a big deal